modules/index/src/main/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskController.java

```diff
@@ -153,46 +159,84 @@ private void onIndexRemoved(RemoveIndexEventParameters parameters) {
 
     private void onPrimaryReplicaElected(PrimaryReplicaEventParameters parameters) {
         inBusyLock(busyLock, () -> {
-            TablePartitionId primaryReplicaId = (TablePartitionId) parameters.groupId();
+
+            PartitionGroupId primaryReplicaId = (PartitionGroupId) parameters.groupId();
 
             if (primaryReplicaId.partitionId() != 0) {
                 // We are only interested in the 0 partition.
                 return;
             }
 
-            int tableId = primaryReplicaId.tableId();
-
             if (isLocalNode(clusterService, parameters.leaseholderId())) {
-                if (localNodeIsPrimaryReplicaForTableIds.add(tableId)) {
-                    scheduleTasksOnPrimaryReplicaElectedBusy(tableId);
-                }
+                scheduleTasksOnPrimaryReplicaElectedBusy(primaryReplicaId);
             } else {
-                if (localNodeIsPrimaryReplicaForTableIds.remove(tableId)) {
-                    changeIndexStatusTaskScheduler.stopTasksForTable(tableId);
-                }
+                scheduleStopTasksOnPrimaryReplicaElected(primaryReplicaId);
             }
         });
     }
 
-    private void scheduleTasksOnPrimaryReplicaElectedBusy(int tableId) {
+    private void scheduleTasksOnPrimaryReplicaElectedBusy(PartitionGroupId partitionGroupId) {
         // It is safe to get the latest version of the catalog because the PRIMARY_REPLICA_ELECTED event is handled on the metastore thread.
         Catalog catalog = catalogService.catalog(catalogService.latestCatalogVersion());
 
-        for (CatalogIndexDescriptor indexDescriptor : catalog.indexes(tableId)) {
-            switch (indexDescriptor.status()) {
-                case REGISTERED:
-                    changeIndexStatusTaskScheduler.scheduleStartBuildingTask(indexDescriptor);
+        IntArrayList tableIds = getTableIdsForPrimaryReplicaElected(catalog, partitionGroupId);
 
-                    break;
+        localNodeIsPrimaryReplicaForTableIds.addAll(tableIds);
 
-                case STOPPING:
-                    changeIndexStatusTaskScheduler.scheduleRemoveIndexTask(indexDescriptor);
+        IntListIterator tableIdsIterator = tableIds.iterator();
+        while (tableIdsIterator.hasNext()) {
```

Comment:
I think a regular loop looks more natural and familiar than an iterator, especially in this case.

Reply:
It requires additional boxing while using a standard iterator. So that is the reason for using `IntListIterator`

Reply:
I agree with Slava, avoiding unnecessary boxing here is a good thing, just as everywhere else. `forEach` looks nice to me